### PR TITLE
IDEA-284073 add support immutability of toList

### PR DESCRIPTION
**Pull request by pyltsin**
I tried to add new logic for "Stream.toList"

https://youtrack.jetbrains.com/issue/IDEA-284073

I hope I did it correctly,
If you find any mistakes I will be happy to fix them.

### CODE REVIEW

File: java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/Mutability.java [hunk 1]
```diff
@@ -64,6 +64,8 @@
   public static final @NotNull String UNMODIFIABLE_VIEW_ANNOTATION = UNMODIFIABLE_VIEW.myAnnotation;
   private static final @NotNull CallMatcher STREAM_COLLECT = CallMatcher.instanceCall(
     CommonClassNames.JAVA_UTIL_STREAM_STREAM, "collect").parameterTypes("java.util.stream.Collector");
+  private static final @NotNull CallMatcher STREAM_COLLECT_TO_LIST = CallMatcher.instanceCall(
+    CommonClassNames.JAVA_UTIL_STREAM_STREAM, "toList");
```
Comment (amaembo) [inline java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/Mutability.java:68]: I would add `.withLanguageLevelAtLeast(JDK_16)`. This matcher can match `toList()` method declared in inheritors, even if `Stream.toList()` is absent. As before JDK_16 there were no `toList()` in the Stream, inheritors were free to assign any other semantics to the result of `toList()`, including the modifiable list. This was true for older versions of `StreamEx` library and probably true for `jOOL` library (haven't checked). So if the project was not updated to Java 16 and uses one of these libraries, the libraries don't violate the contract there but we will have incorrect warnings.

File: java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/java/inliner/StreamChainInliner.java [hunk 2]
```diff
@@ -69,6 +69,7 @@
     instanceCall(JAVA_UTIL_STREAM_STREAM, "reduce").parameterTypes("T", "java.util.function.BinaryOperator");
   private static final CallMatcher COLLECT_TERMINAL =
     instanceCall(JAVA_UTIL_STREAM_STREAM, "collect").parameterTypes("java.util.stream.Collector");
+  private static final CallMatcher COLLECT_TO_LIST_TERMINAL = instanceCall(JAVA_UTIL_STREAM_STREAM, "toList");
```
Comment (amaembo) [inline java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/java/inliner/StreamChainInliner.java:72]: Please rename it to `TO_LIST_TERMINAL`. Naming here assumes mentioning of called methods but you are not calling `collect()` here.

File: java/java-tests/testSrc/com/intellij/java/codeInspection/DataFlowInspection16WithMockTest.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package com.intellij.java.codeInspection;
+
+import com.intellij.JavaTestUtil;
+import com.intellij.testFramework.LightProjectDescriptor;
+import org.jetbrains.annotations.NotNull;
+
+public class DataFlowInspection16WithMockTest extends DataFlowInspectionTestCase {
```
Comment (amaembo) [inline java/java-tests/testSrc/com/intellij/java/codeInspection/DataFlowInspection16WithMockTest.java:8]: I think it should be ok to add a new test to the existing `DataFlowInspection16Test` or `DataFlowInspection17Test` class. It's not so important that the particular test has library class mocks, as this doesn't affect other tests in the same TestCase class and doesn't require any specific setup steps. Also, I think that nothing bad will happen if mockJDK already has the Stream class and you redefine it inside the test. The references will be naturally resolved to the redefined one, as it's in the same file. In this case, you probably don't need to mock BaseStream at all.

**Review by amaembo — CHANGES_REQUESTED** [review]

File: java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/Mutability.java (hunk 1 above)
Comment (pyltsin) [reply java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/Mutability.java]: Thank you for your explanation. Fixed it

File: java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/java/inliner/StreamChainInliner.java (hunk 2 above)
Comment (pyltsin) [reply java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/java/inliner/StreamChainInliner.java]: Renamed

File: java/java-tests/testSrc/com/intellij/java/codeInspection/DataFlowInspection16WithMockTest.java (hunk 3 above)
Comment (pyltsin) [reply java/java-tests/testSrc/com/intellij/java/codeInspection/DataFlowInspection16WithMockTest.java]: I moved test and created mock in this test-case.

Thanks again!

**Comment by pyltsin** [comment]
I tried to fix. Thank you!

**Comment by amaembo** [comment]
Pushed to internal repo, thanks.

**Pull request closed without merging**